### PR TITLE
chore(graphql api): Aggregate component metrics & component refactor

### LINKED ISSUE #5319: Negative throughput from `vector top`
### Vector Version

```
vector 0.11.0 (gafc64e5 x86_64-unknown-linux-gnu 2020-12-01)
```

### Vector Configuration File

```toml
# Data dir is location controlled at the `DaemonSet`.
data_dir = "/vector-data-dir"

[api]
enabled = true

[log_schema]
host_key = "host"
message_key = "message"
source_type_key = "source_type"
timestamp_key = "timestamp"

[sources.internal_metrics]
type = "internal_metrics"

# Ingest logs from Kubernetes.
[sources.kubernetes_logs]
type = "kubernetes_logs"

[sources.json_logs]
type = "file"
include = ["/var/log/kube-audit/*.json", "/var/lib/rancher/rke/log/*.log"]

[sources.system_logs]
type = "file"
include = ["/var/log/syslog*", "/var/log/auth.log*"]

[transforms.json_parser]
type = "json_parser"
inputs = ["json_logs"]
drop_invalid = true

[sinks.vector]
type = "vector"
inputs = ["kubernetes_logs","json_parser","system_logs"]
address = "vector-aggregator:9000"

[sinks.prometheus]
type = "prometheus_exporter"
inputs = ["internal_metrics"]
address = "0.0.0.0:9598"
default_namespace = "agent"
```

### Expected Behavior

Greater than 0 throughput

### Actual Behavior

image

Negative throughput reported

I'm going to try some tests and see if it happens on other sources aside from `file`

### PR DESCRIPTION
**Pull request by leebenson**
Attempts to fix #5319 by aggregating component metrics.

@spencergilbert, would you mind giving this a go to see if this fixes the negative file source stats in `vector top`? Thanks!

Also refactors component schema layout, in anticipation of expanding schema to accommodate component-specific meta per #5402.

### CODE REVIEW

File: src/api/schema/metrics/mod.rs [hunk 1]
```diff
@@ -253,32 +250,40 @@ fn metrics_sorted(interval: i32) -> impl Stream<Item = Vec<Metric>> {
...
+            if m.name.as_str().eq("processed_events_total")
+                && m.tag_matches("component_name", &component_name) =>
+        {
+            Some(m)
+        }
+        _ => None,
+    });
+    let m = metrics.next()?;
+
+    Some(ProcessedEventsTotal::new(metrics.fold(m, |mut m1, m2| {
```
Comment (leebenson) [inline src/api/schema/metrics/mod.rs:264]: This is the main change in the PR. Most everything else is just component refactoring.

**Comment by spencergilbert** [comment]
> Attempts to fix #5319 by aggregating component metrics.
> 
> @spencergilbert, would you mind giving this a go to see if this fixes the negative file source stats in `vector top`? Thanks!
> 
> Also refactors component schema layout, in anticipation of expanding schema to accommodate component-specific meta per #5402.

Can just run client side right?

**Comment by leebenson** [comment]
> Can just run client side right?

This is an API change in how metrics are sampled (aggregated, instead of stopping at the first matching metric), so it'd be require swapping the Vector instance itself.

No worries if that's not practical for you right now, we can test internally.

**Comment by spencergilbert** [comment]
> > Can just run client side right?
> 
> This is an API change in how metrics are sampled (aggregated, instead of stopping at the first matching metric), so it'd be require swapping the Vector instance itself.
> 
> No worries if that's not practical for you right now, we can test internally.

Gotcha, wasn't sure based on my quick reading of the code. I can easily test if there's a docker image to pull, otherwise it's a bit more of a hassle.

**Comment by leebenson** [comment]
Closing this, as I have a branch that supersedes this and confirms a fix for reporting file metrics.

**Pull request closed without merging**